Onesixtyone inventory plugin (Fixes #41443)

victorock opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Add inventory plugin for _onesixtyone_, which is lightweight and fast in discovery of devices running SNMP.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
Fixes #41443 

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
onesixtyone

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below
ansible-inventory -i inventory --list
{
    "_meta": {
        "hostvars": {
            "ip-172-16-96-168.eu-central-1.compute.internal": {
                "ansible_host": "172.16.96.168", 
                "ansible_network_os": "ios", 
                "snmp_community": "public", 
                "snmp_sysdescr": "Cisco IOS Software [Gibraltar], Virtual XE Software (X86_64_LINUX_IOSD-UNIVERSALK9-M), Version 16.10.1b, RELEASE SOFTWARE (fc1)  Technical Support: http://www.cisco.com/techsupport  Copyright (c) 1986-2018 by Cisco Systems, Inc.  Compiled Wed 12-Dec-18 02"
            }
        }
    }, 
    "all": {
        "children": [
            "ios", 
            "ungrouped"
        ]
    }, 
    "ios": {
        "hosts": [
            "ip-172-16-96-168.eu-central-1.compute.internal"
        ]
    }, 
    "ungrouped": {}
}
```

Commit:
onesixtyone:
- inventory plugin using onesixtyone  to discover devices using snmp.
- group together hosts with the same platforms.

Commit:
Typo of platform. It is a dictionary and not a list.

Akasurde commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,203 @@
+DOCUMENTATION = '''
+    name: onesixtyone
+    plugin_type: inventory
+    version_added: "2.9"
```
```suggestion
    version_added: "2.8"
```

Akasurde commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,203 @@
+# Copyright (c) 2019, Victor da Costa <vdacosta@redhat.com>
+# Copyright (c) 2019 Ansible Project
```
```suggestion
# Copyright (c) 2019, Ansible Project
```

Akasurde commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,203 @@
+    name: onesixtyone
+    plugin_type: inventory
+    version_added: "2.9"
+    short_description: Uses onesityone to find hosts to target
```
```suggestion
    short_description: Uses onesixtyone to find hosts to target
```

Akasurde commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,203 @@
+    version_added: "2.9"
+    short_description: Uses onesityone to find hosts to target
+    description:
+        - Uses a YAML configuration file with a valid YAML extension.
```
```suggestion
        - Uses any YAML file which ends with onesixtyone.yml or onesixtyone.yaml as a configuration file.
```

Akasurde commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,203 @@
+  import gevent
+  from gevent import socket
+except:
+  raise Exception('this inventory plugin requires the gevent library. Try: yum install python-gevent')
```
```suggestion
  raise AnsibleError('This inventory plugin requires the gevent Python library. Please install using pip or respective package manager.')
```

victorock replied:
I am using AnsibleParserError exception hence it seems more associated to InventoryModules.

Akasurde commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,203 @@
+    def __init__(self):
+
+        self._onesixtyone = None
+        for path in os.environ.get('PATH').split(':'):
```
I think we can use - `from ansible.module_utils.common.process import get_bin_path`

webknjaz replied:
s/can/should/

victorock replied:
```PYTHON
    def __init__(self):
        super(InventoryModule, self).__init__()
        self._onesixtyone = get_bin_path(self.NAME, True)
```

victorock replied:
Cannot use get_bin_path, hence it is in ansible.module_utils.basic as method of class AnsibleModule...

victorock replied:
Added get_bin_path as staticmethod in InventoryModule class of my inventory plugin.

bcoca replied:
there is 'non AnsibleModule' version in  lib/ansible/module_utils/common/process

webknjaz commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,203 @@
+    def verify_file(self, path):
+
+        if super(InventoryModule, self).verify_file(path):
+          for fn in ('onesixtyone'):
```
Why would you do something like this?

victorock replied:
Was with the idea to have multiple files.
Moving to:
```PYTHON
    def verify_file(self, path):

        if not super(InventoryModule, self).verify_file(path):
            return False

        for suffix in ('yaml', 'yml'):
            maybe = '{fn}.{suffix}'.format(fn=self.NAME, suffix=suffix)
            if path.endswith(maybe):
                return True
```

webknjaz commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,203 @@
+
+    def verify_file(self, path):
+
+        if super(InventoryModule, self).verify_file(path):
```
Use "guard expression" style here to reduce nesting
```suggestion
        if not super(InventoryModule, self).verify_file(path):
            return False
```

victorock replied:
```PYTHON
    def verify_file(self, path):

        if not super(InventoryModule, self).verify_file(path):
            return False

        for suffix in ('yaml', 'yml'):
            maybe = '{fn}.{suffix}'.format(fn=self.NAME, suffix=suffix)
            if path.endswith(maybe):
                return True
```

webknjaz commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,203 @@
+        if super(InventoryModule, self).verify_file(path):
+          for fn in ('onesixtyone'):
+              for suffix in ('yaml', 'yml'):
+                  maybe = '{fn}.{suffix}'.format(fn=fn, suffix=suffix)
```
```suggestion
                  maybe = '{fn}.{suffix}'.format(fn=self.NAME, suffix=suffix)
```

victorock replied:
yes, that is what i did at the end...

webknjaz commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,203 @@
+            raise AnsibleParserError("failed to parse %s: %s " % (to_native(path), to_native(e)))
+
+    def _match_platform(self, text):
+        ''' Matches platforms against text in sysDescr
```
Please follow the guidance of PEP257 to write nice docstrings

Commit:
Changelog:
- pep8 fixes
- netaddr for cidrs
- documentation as per PEP257
- comments and suggestions

Commit:
get_bin_path

Commit:
Fix pep8 E722

bcoca commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,236 @@
+try:
+    from netaddr import IPSet
+except Exception:
+    raise AnsibleParserError('this inventory plugin requires the netapp library. Try: pip install netaddr')
```
please check developer docs for how to deal with required libs, this should not give an exception at this point (load) , it should happen at `parse()`

victorock replied:
```PYTHON
    def parse(self, inventory, loader, path, cache=False):
        super(InventoryModule, self).parse(inventory, loader, path, cache=cache)
        self._onesixtyone = get_bin_path(self.NAME, True)
        self._read_config_data(path)

        if not HAS_NETADDR:
            raise AnsibleParserError('this inventory plugin requires the netapp library. Try: pip install netaddr')
        
        if not HAS_GEVENT:
            raise AnsibleParserError('this inventory plugin requires the gevent library. Try: yum install python-gevent')
```

bcoca commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,236 @@
+
+    def __init__(self):
+        super(InventoryModule, self).__init__()
+        self._onesixtyone = get_bin_path(self.NAME, True)
```
same here, this will cause exception on load, not on parse

victorock replied:
```PYTHON
    def parse(self, inventory, loader, path, cache=False):
        super(InventoryModule, self).parse(inventory, loader, path, cache=cache)
        self._onesixtyone = get_bin_path(self.NAME, True)
        self._read_config_data(path)

        if not HAS_NETADDR:
            raise AnsibleParserError('this inventory plugin requires the netapp library. Try: pip install netaddr')
        
        if not HAS_GEVENT:
            raise AnsibleParserError('this inventory plugin requires the gevent library. Try: yum install python-gevent')
```

Commit:
Exception raised from InventoryModule.
get_bin_path call from InventoryModule.parse()

Commit:
pep8 fixes

Commit:
get_bin_path as staticmethod in InventoryModule

Commit:
from ansible.module_utils.common.process import get_bin_path

Commit:
remove unecessary import

victorock commented:
@bcoca @Akasurde @webknjaz done

Akasurde commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,244 @@
+          required: false
+    notes:
+      - Only discover devices running SNMPv2.
+      - TODO: Add caching support.
```
```suggestion
      - 'TODO: Add caching support.'
```

Akasurde commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,244 @@
+            raise AnsibleParserError('this inventory plugin requires the gevent library. Try: yum install python-gevent')
+
+        try:
+            # Using gevent to pallelize reverse dns lookups and adding entries to inventory
```
```suggestion
            # Using gevent to parallelize reverse dns lookups and adding entries to inventory
```

Akasurde committed:
typo in comments

Co-Authored-By: victorock <victorockeiro@gmail.com>

Akasurde committed:
typo in parallelize

Co-Authored-By: victorock <victorockeiro@gmail.com>

victorock commented:
@bcoca @Akasurde any additional requirement to fix?

bcoca commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,244 @@
+
+    def parse(self, inventory, loader, path, cache=False):
+        super(InventoryModule, self).parse(inventory, loader, path, cache=cache)
+        self._onesixtyone = get_bin_path(self.NAME, True)
```
this produces an exception when binary cannot be found, you should capture and return parser error

victorock replied:
Moved it under the `try` in `parser` method.

bcoca commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,244 @@
+    def parse(self, inventory, loader, path, cache=False):
+        super(InventoryModule, self).parse(inventory, loader, path, cache=cache)
+        self._onesixtyone = get_bin_path(self.NAME, True)
+        self._read_config_data(path)
```
why read the data before you verify you have your requirements?

victorock replied:
Moved it under the `try` in `parser method`.

bcoca commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,244 @@
+            jobs = [gevent.spawn(self._add_device, self, device) for device in self._discovery()]
+            gevent.joinall(jobs, timeout=20)
+
+        except Exception as e:
```
instead of a catchall here, you should give more relevant errors in _discovery and  _add_device, which you already do in some cases, but then this catchall obscures

bcoca commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,244 @@
+    HAS_GEVENT = False
+
+
+class InventoryModule(BaseInventoryPlugin, Constructable, Cacheable):
```
you subclass constructable and cacheable but don't use them?

victorock replied:
The initial plan was to use them...removed `Contructable` and `Cacheable`.

webknjaz commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,244 @@
+        self._onesixtyone = get_bin_path(self.NAME, True)
+        self._read_config_data(path)
+
+        if not HAS_NETADDR:
```
you should probably put these check even before `super()` line

victorock replied:
The idea was to have a consistent way to :
* Easily identify if superclass methods are being called/initialize or not.
* Avoid any unexpected behaviours due subclass dependency of whatever is being done by superclass method...

Moved checks to be the first thing happening in `parse method`.

webknjaz commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,244 @@
+            raise AnsibleParserError('this inventory plugin requires the netaddr library. Try: pip install netaddr')
+
+        if not HAS_GEVENT:
+            raise AnsibleParserError('this inventory plugin requires the gevent library. Try: yum install python-gevent')
```
why it's `yum` here, but `pip` in the previous error? be consistent and tell users to use what they have. which can be many more package manager options

victorock replied:
Doing consistency on `Try: pip`....

webknjaz commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,244 @@
+
+        try:
+            # Using gevent to parallelize reverse dns lookups and adding entries to inventory
+            jobs = [gevent.spawn(self._add_device, self, device) for device in self._discovery()]
```
this line will probably not generate an exception and should be outside of try/except block

bcoca replied:
both _discovery and _add_device can throw exceptions, gevent SHOULD push them to the join below, but that is not always the case and depends on timing (they solved this problem a while ago, but we are not specifying version requirement and would have to deal with least common denominator)

victorock replied:
As per your comments, i am doing the following:
* Keeping `gevent` under `try`
* Adding `catch` for gevent `exceptions` (`LoopExit`).
* Removing `catch-all` so `AnsibleParserError` exceptions cascade up the hill...

webknjaz commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,244 @@
+        text -- sysdescr text
+        """
+        platforms = {
+            'asa': re.compile(r'^Cisco\sAdaptive\sSecurity\sAppliance'),
```
No need to compile explicitly, `re` will hold compiled cache internally anyway.

victorock replied:
I prefer to keep the `compile`, method, to explicitly know what is happening. Do you want me to change from `re.compile(r'Cisco\sAdaptive\sSecurity\sAppliance')` to `r'Cisco\sAdaptive\sSecurity\sAppliance'` ?

victorock replied:
@webknjaz any comments?

webknjaz commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,244 @@
+        """
+        platforms = {
+            'asa': re.compile(r'^Cisco\sAdaptive\sSecurity\sAppliance'),
+            'ios': re.compile(r'^Cisco\sIOS\sSoftware'),
```
You use `re.match` which already checks the beginning of a string. So caret `^` is not needed.

https://docs.python.org/3/library/re.html#search-vs-match

victorock replied:
Removed the `ˆ` from the beginning of the `regex`, even though i'd keep to remember that we are checking the `beginning of a string`

webknjaz commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,244 @@
+            'dellos': re.compile(r'^Dell\sApplication\sSoftware'),
+            'junos': re.compile(r'^JUNOS'),
+            'aruba': re.compile(r'^Aruba\sOperating\sSystem\sSoftware'),
+            'eos': re.compile(r'^Arista')
```
I prefer to have a trailing comma everywhere, it makes further diff review easier.
```suggestion
            'eos': re.compile(r'^Arista'),
```

victorock replied:
Added ','

victorock committed:
fix as per suggestions

- organize errors/exceptions
- add trailing ','
- remove '^' from regex

bcoca commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -74,12 +74,13 @@
 
 
-class InventoryModule(BaseInventoryPlugin, Constructable, Cacheable):
+class InventoryModule(BaseInventoryPlugin):
```
i would advise to keep constructable and just implement the functionality, it allows the user to create groups/variables via  compose so you don't have to hardcode it into the plugin

victorock replied:
@bcoca it *cannot* be `constructable` because i am only able to `group_by` `platform`....

bcoca replied:
not sure i understand that, won't users also want to group by other data?

victorock replied:
Nope, the only relevant data is the one i am using, which is the platform itself, so additional interaction to grab more details from the device is possible through `<platform>_facts`, and grouping them with `group_by`.

bcoca replied:
That Requries a 2nd interaction and playbook hoops, some of the data is already available as output from onesixtyone, why not make it available to user instead of immediately discarding it?

victorock committed:
fix pep8

victorock commented:
@bcoca @Akasurde hopefully everything is good now.

victorock commented:
@Akasurde @bcoca any news?

webknjaz commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,251 @@
+DOCUMENTATION = '''
+    name: onesixtyone
+    plugin_type: inventory
+    version_added: "2.8"
```
```suggestion
    version_added: "2.9"
```

webknjaz commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,251 @@
+try:
+    from netaddr import IPSet
+    HAS_NETADDR = True
+except Exception:
```
ImportError?

webknjaz commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,251 @@
+    NAME = 'onesixtyone'
+
+    def __init__(self):
+        super(InventoryModule, self).__init__()
```
Empty proxy-init doesn't make sense and should be removed.

webknjaz commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,251 @@
+        # Find the onesixtyone binary
+        try:
+            self._onesixtyone = get_bin_path(self.NAME, True)
+        except Exception:
```
Please, use actual exception which may happen. Wholesale catch-all is bad.

webknjaz commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,251 @@
+        }
+        platforms.update(self.get_option('platform'))
+
+        for platform, rule in platforms.items():
```
```suggestion
        return next(platform for platform, rule in platforms.items() if rule.match(text), 'unknown')
```

webknjaz commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,251 @@
+            communities_tempfile.flush()
+
+            # add command-line option -c
+            cmd.append('-c')
```
It'd be probably better readable with
```suggestion
            cmd.extend(('-c', communities_tempfile.name))
```
instead

webknjaz commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,251 @@
+
+                # execute
+                try:
+                    self.display.vvv("running command: {cmd}".format(cmd=cmd_string))
```
This should go outside of try block

webknjaz commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,251 @@
+                matches = None
+
+                for line in cmd_output.splitlines():
+                    self.display.vvv("parsing output line: {line}".format(line=line))
```
this should probably go to the debug level

webknjaz commented on lib/ansible/plugins/inventory/onesixtyone.py:
```diff
@@ -0,0 +1,251 @@
+            if host.endswith('.in-addr.arpa'):
+                obj.display.vvv("Reverse DNS for host {host} contains 'in-addr.arpa', using {ansible_host}`".format(host=host, ansible_host=ansible_host))
+                host = ansible_host
+        except Exception:
```
no wholesale exceptions plz

Akasurde commented:
Hi @victorock, Thank you very much for your interest in Ansible. This plugin/module is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.general<br><br>If you have further questions please stop by IRC or the mailing list:<br><br>* IRC: #ansible on irc.freenode.net<br>* mailing list: https://groups.google.com/forum/#!forum/ansible-project<br><br>needs_info<br>
